csharp/src/Ice/InstrumentationI.cs

```diff
@@ -23,1072 +22,748 @@ public InvocationMetrics()
+                Add("adapterName", obj => (obj as ConnectionHelper)?.ConnectionInfo.AdapterName);
+                Add("connectionId", obj => (obj as ConnectionHelper)?.ConnectionInfo.ConnectionId);
 
-        public override void InitMetrics(DispatchMetrics v) => v.Size += _size;
+                Add("localHost", obj => (obj as ConnectionHelper)?.IPConnectionInfo?.LocalAddress);
+                Add("localPort", obj => (obj as ConnectionHelper)?.IPConnectionInfo?.LocalPort);
+                Add("remoteHost", obj => (obj as ConnectionHelper)?.IPConnectionInfo?.RemoteAddress);
+                Add("remotePort", obj => (obj as ConnectionHelper)?.IPConnectionInfo?.RemotePort);
 
-        public string GetMode() => _current.RequestId == 0 ? "oneway" : "twoway";
+                Add("mcastHost",
+                    obj => ((obj as ConnectionHelper)?.ConnectionInfo as UDPConnectionInfo)?.McastAddress);
+                Add("mcastPort",
+                    obj => ((obj as ConnectionHelper)?.ConnectionInfo as UDPConnectionInfo)?.McastPort);
 
-        public string GetId()
-        {
-            if (_id == null)
-            {
-                var os = new StringBuilder();
-                if (_current.Identity.Category != null && _current.Identity.Category.Length > 0)
-                {
-                    os.Append(_current.Identity.Category).Append('/');
-                }
-                os.Append(_current.Identity.Name).Append(" [").Append(_current.Operation).Append(']');
-                _id = os.ToString();
+                Add("endpoint", obj => (obj as ConnectionHelper)?.Endpoint);
+                Add("endpointType", obj => (obj as ConnectionHelper)?.Endpoint?.Type);
+                Add("endpointIsDatagram", obj => (obj as ConnectionHelper)?.Endpoint?.IsDatagram);
+                Add("endpointIsSecure", obj => (obj as ConnectionHelper)?.Endpoint?.IsSecure);
+                Add("endpointTimeout", obj => (obj as ConnectionHelper)?.Endpoint?.Timeout);
+                Add("endpointCompress", obj => (obj as ConnectionHelper)?.Endpoint?.HasCompressionFlag);
+                Add("endpointHost", obj => ((obj as ConnectionHelper)?.Endpoint as IPEndpoint)?.Host);
+                Add("endpointPort", obj => ((obj as ConnectionHelper)?.Endpoint as IPEndpoint)?.Port);
             }
-            return _id;
         }
+    }
 
-        public string GetParent() => _current.Adapter.Name;
+    internal class ConnectionObserver
+        : ObserverWithDelegate<ConnectionMetrics, IConnectionObserver>, IConnectionObserver
+    {
+        private int _receivedBytes;
+        private int _sentBytes;
 
-        public ConnectionInfo? GetConnectionInfo()
+        public void ReceivedBytes(int num)
         {
-            if (_current.Connection != null)
-            {
-                return _current.Connection.GetConnectionInfo();
-            }
-            return null;
+            _receivedBytes = num;
+            ForEach(ReceivedBytesUpdate);
```

Comment:
Can't we just use a lambda instead of a private method here like we do in other places?